File: app/src/main/java/org/wikipedia/feed/announcement/AnnouncementCardView.java
package org.wikipedia.feed.announcement;

import android.content.Context;
import android.net.Uri;
import android.text.TextUtils;
import android.text.method.LinkMovementMethod;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.wikipedia.R;
import org.wikipedia.feed.model.Card;
import org.wikipedia.feed.view.DefaultFeedCardView;
import org.wikipedia.richtext.RichTextUtil;
import org.wikipedia.util.DimenUtil;
import org.wikipedia.util.StringUtil;
import org.wikipedia.views.FaceAndColorDetectImageView;
import org.wikipedia.views.WikiCardView;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

import static android.view.ViewGroup.LayoutParams.MATCH_PARENT;

public class AnnouncementCardView extends DefaultFeedCardView<AnnouncementCard> {
    public interface Callback {
        void onAnnouncementPositiveAction(@NonNull Card card, @NonNull Uri uri);
        void onAnnouncementNegativeAction(@NonNull Card card);
    }

    @BindView(R.id.view_announcement_container) WikiCardView container;
    @BindView(R.id.view_announcement_header_image) FaceAndColorDetectImageView headerImageView;
    @BindView(R.id.view_announcement_text) TextView textView;
    @BindView(R.id.view_announcement_action_positive) Button actionViewPositive;
    @BindView(R.id.view_announcement_action_negative) Button actionViewNegative;
    @BindView(R.id.view_announcement_action_positive_dialog) Button actionViewPositiveDialog;
    @BindView(R.id.view_announcement_action_negative_dialog) Button actionViewNegativeDialog;
    @BindView(R.id.view_announcement_card_buttons_container) View actionViewContainer;
    @BindView(R.id.view_announcement_card_dialog_buttons_container) View actionViewDialogsContainer;
    @BindView(R.id.view_announcement_footer_text) TextView footerTextView;
    @BindView(R.id.view_announcement_footer_border) View footerBorderView;
    @Nullable private Callback callback;

    public AnnouncementCardView(@NonNull Context context) {
        super(context);
        inflate(context, R.layout.view_card_announcement, this);
        ButterKnife.bind(this);

        textView.setMovementMethod(LinkMovementMethod.getInstance());
        footerTextView.setMovementMethod(LinkMovementMethod.getInstance());
    }

    @Override public void setCard(@NonNull AnnouncementCard card) {
        super.setCard(card);

        if (!TextUtils.isEmpty(card.extract())) {
            textView.setText(StringUtil.fromHtml(card.extract()));
        }

        if (!card.hasAction()) {
            actionViewContainer.setVisibility(GONE);
        } else {
            actionViewContainer.setVisibility(VISIBLE);
            actionViewPositive.setText(card.actionTitle());
            actionViewPositiveDialog.setText(card.actionTitle());
        }
        if (!TextUtils.isEmpty(card.negativeText())) {
            actionViewNegative.setText(card.negativeText());
            actionViewNegativeDialog.setText(card.negativeText());
        } else {
            actionViewNegative.setVisibility(GONE);
            actionViewNegativeDialog.setVisibility(GONE);
        }

        if (card.hasImage()) {
            headerImageView.setVisibility(VISIBLE);
            headerImageView.loadImage(card.image());
        } else {
            headerImageView.setVisibility(GONE);
        }

        if (card.imageHeight() > 0) {
            ViewGroup.MarginLayoutParams params = (ViewGroup.MarginLayoutParams) headerImageView.getLayoutParams();
            params.height = DimenUtil.roundedDpToPx(card.imageHeight());
            headerImageView.setLayoutParams(params);
        }

        if (card.hasFooterCaption()) {
            footerTextView.setText(StringUtil.fromHtml(card.footerCaption()));
            RichTextUtil.removeUnderlinesFromLinks(footerTextView);
        } else {
            footerTextView.setVisibility(GONE);
            footerBorderView.setVisibility(GONE);
        }

        if (card.hasBorder()) {
            container.setStrokeColor(getResources().getColor(R.color.red30));
            container.setStrokeWidth(10);
        } else {
            container.setDefaultBorder();
        }

        if (card.isArticlePlacement()) {
            actionViewContainer.setVisibility(GONE);
            actionViewDialogsContainer.setVisibility(VISIBLE);
            container.setLayoutParams(new LinearLayout.LayoutParams(MATCH_PARENT, MATCH_PARENT));
            container.setRadius(0);
        }
    }

    @OnClick({R.id.view_announcement_action_positive, R.id.view_announcement_action_positive_dialog})
    void onPositiveActionClick() {
        if (getCard() != null) {
            if (getCallback() != null) {
                getCallback().onAnnouncementPositiveAction(getCard(), getCard().actionUri());
            } else if (callback != null) {
                callback.onAnnouncementPositiveAction(getCard(), getCard().actionUri());
            }
        }
    }

    @OnClick({R.id.view_announcement_action_negative, R.id.view_announcement_action_negative_dialog})
    void onNegativeActionClick() {
        if (getCard() != null) {
            if (getCallback() != null) {
                getCallback().onAnnouncementNegativeAction(getCard());
            } else if (callback != null) {
                callback.onAnnouncementNegativeAction(getCard());
            }
        }
    }

    public void setCallback(@NonNull Callback callback) {
        this.callback = callback;
    }
}

File: app/src/main/java/org/wikipedia/feed/announcement/AnnouncementCardView.kt
package org.wikipedia.feed.announcement

import android.app.Activity
import android.content.Context
import android.net.Uri
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.ImageView
import androidx.core.content.ContextCompat
import androidx.core.text.method.LinkMovementMethodCompat
import androidx.core.view.updateLayoutParams
import org.wikipedia.R
import org.wikipedia.analytics.eventplatform.BreadCrumbLogEvent
import org.wikipedia.databinding.ViewCardAnnouncementBinding
import org.wikipedia.feed.model.Card
import org.wikipedia.feed.onboarding.YIROnboardingCard
import org.wikipedia.feed.view.DefaultFeedCardView
import org.wikipedia.util.DimenUtil
import org.wikipedia.util.StringUtil

class AnnouncementCardView(context: Context) : DefaultFeedCardView<AnnouncementCard>(context) {
    interface Callback {
        fun onAnnouncementPositiveAction(card: Card, uri: Uri)
        fun onAnnouncementNegativeAction(card: Card)
    }

    private val binding = ViewCardAnnouncementBinding.inflate(LayoutInflater.from(context), this, true)
    var localCallback: Callback? = null

    init {
        binding.viewAnnouncementText.movementMethod = LinkMovementMethodCompat.getInstance()
        binding.viewAnnouncementFooterText.movementMethod = LinkMovementMethodCompat.getInstance()
        binding.viewAnnouncementActionPositive.setOnClickListener { onPositiveActionClick() }
        binding.viewAnnouncementDialogActionPositive.setOnClickListener { onPositiveActionClick() }
        binding.viewAnnouncementActionNegative.setOnClickListener { onNegativeActionClick() }
        binding.viewAnnouncementDialogActionNegative.setOnClickListener { onNegativeActionClick() }
    }

    override var card: AnnouncementCard? = null
        set(value) {
            field = value
            value?.let {
                if (!it.extract().isNullOrEmpty()) {
                    binding.viewAnnouncementText.text = StringUtil.fromHtml(it.extract())
                }
                if (!it.hasAction()) {
                    binding.viewAnnouncementCardButtonsContainer.visibility = GONE
                } else {
                    binding.viewAnnouncementCardButtonsContainer.visibility = VISIBLE
                    binding.viewAnnouncementActionPositive.text = it.actionTitle()
                    binding.viewAnnouncementActionPositive.tag = it.getId()
                    binding.viewAnnouncementDialogActionPositive.text = it.actionTitle()
                    binding.viewAnnouncementDialogActionPositive.tag = it.getId()
                }
                if (!it.negativeText().isNullOrEmpty()) {
                    binding.viewAnnouncementActionNegative.text = it.negativeText()
                    binding.viewAnnouncementActionNegative.tag = it.getId()
                    binding.viewAnnouncementDialogActionNegative.text = it.negativeText()
                    binding.viewAnnouncementDialogActionNegative.tag = it.getId()
                } else {
                    binding.viewAnnouncementActionNegative.visibility = GONE
                    binding.viewAnnouncementDialogActionNegative.visibility = GONE
                }
                if (it.hasImage()) {
                    binding.viewAnnouncementHeaderImage.visibility = VISIBLE
                    if (it.aspectRatio() != 0.0) {
                        binding.viewAnnouncementHeaderImage.scaleType = ImageView.ScaleType.FIT_CENTER
                        binding.viewAnnouncementHeaderImage.post {
                            if ((context as? Activity)?.isDestroyed == true) {
                                return@post
                            }
                            binding.viewAnnouncementHeaderImage.updateLayoutParams {
                                height = (binding.viewAnnouncementHeaderImage.width / it.aspectRatio()).toInt()
                            }
                            binding.viewAnnouncementHeaderImage.loadImage(it.image())
                        }
                    } else {
                        binding.viewAnnouncementHeaderImage.loadImage(it.image())
                    }
                } else {
                    binding.viewAnnouncementHeaderImage.visibility = GONE
                }
                if (it.imageHeight() > 0) {
                    binding.viewAnnouncementHeaderImage.updateLayoutParams<MarginLayoutParams> {
                        height = DimenUtil.roundedDpToPx(it.imageHeight().toFloat())
                    }
                }
                if (it.hasFooterCaption()) {
                    binding.viewAnnouncementFooterText.text = StringUtil.fromHtml(it.footerCaption())
                } else {
                    binding.viewAnnouncementFooterText.visibility = GONE
                    binding.viewAnnouncementFooterBorder.visibility = GONE
                }
                if (it.hasBorder()) {
                    binding.viewAnnouncementContainer.strokeColor = ContextCompat.getColor(context, R.color.red700)
                    binding.viewAnnouncementContainer.strokeWidth = 10
                } else {
                    binding.viewAnnouncementContainer.setDefaultBorder()
                }
                if (it.isArticlePlacement) {
                    binding.viewAnnouncementCardButtonsContainer.visibility = GONE
                    binding.viewAnnouncementCardDialogButtonsContainer.visibility = VISIBLE
                    binding.viewAnnouncementContainer.layoutParams = LayoutParams(
                        ViewGroup.LayoutParams.MATCH_PARENT,
                        ViewGroup.LayoutParams.MATCH_PARENT
                    )
                    binding.viewAnnouncementContainer.radius = 0f
                }
            }

            if (value is YIROnboardingCard) {
                BreadCrumbLogEvent.logImpression(context, "YIR2024")
            }
        }

    private fun onPositiveActionClick() {
        card?.let {
            callback?.onAnnouncementPositiveAction(it, it.actionUri())
            localCallback?.onAnnouncementPositiveAction(it, it.actionUri())
        }
    }

    private fun onNegativeActionClick() {
        card?.let {
            callback?.onAnnouncementNegativeAction(it)
            localCallback?.onAnnouncementNegativeAction(it)
        }
    }
}
